### ISSUE TITLE
Global CSRF Protection

### ISSUE THREAD

**Issue by WGierke** [concept-to-clinic/concept-to-clinic#134]
This issue is a follow-up of the discussion from [here](https://github.com/concept-to-clinic/concept-to-clinic/pull/101#issuecomment-330208795).

[Cross-Site Request Forgery (CSRF) is an attack that forces an end user to execute unwanted actions on a web application in which they're currently authenticated.](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)) As a defensive measure, one configures the server to send a random token to the client which has to be sent back when e.g. a POST request is performed. This ensures that the origin of the POST request really is the own application.
To solve this issue, please send a CSRF token by default to the (Vue.js) client and pass it back when performing POST/PUT/... requests to the server, by default as well.

## Expected Behavior
The server should set a CSRF token per default with every response.
When sending data to the server (e.g. to select the lung orientation as specified in #34), this token should be added automatically to the payload.

## Current Behavior
We're not using CSRF tokens.

## Possible Implementation
Letting Django respond to every request with a CSRF token seems pretty easy by e.g. adding
`'config.middleware.ForceCsrfCookieMiddleware',` to `MIDDLEWARE` and adding 
```
from django.utils.deprecation import MiddlewareMixin
from django.middleware.csrf import get_token

class ForceCsrfCookieMiddleware(MiddlewareMixin):
    """
    Forces the CSRF cookie to be set for each request
    """
    def process_request(self, request):
        get_token(request)
```
in `interface/config/middleware.py`.
[Reference](https://stackoverflow.com/questions/32231607/django-forcing-csrf-token-on-all-responses)

Regarding #125 , letting Vue (with axios) respond with the correct CSRF token seems to be as easy as 
```
import axios from 'axios';
axios.defaults.xsrfCookieName = 'csrftoken'
axios.defaults.xsrfHeaderName = 'X-CSRFToken'
```
[Reference](https://cbuelter.wordpress.com/2017/04/10/django-csrf-with-axios/)

## Acceptance criteria
- [ ] the token is added to every Django response by default
- [ ] the token is added to every payload send to the server by default
- [ ] appropriate tests

**Comment by isms** [comment]
Concur that CSRF is an open question, but before we move forward some thoughts:

### Status quo = we actually are using CSRF tokens

The issue says:

> We're not using CSRF tokens.

... but that's not really true. The backend (final arbiter of any requests) is currently using and will currently insist on CSRF tokens for `PUT`, `POST`, `PATCH` because the `CsrfViewMiddleware` [is installed](https://github.com/concept-to-clinic/concept-to-clinic/blob/master/interface/config/settings/base.py#L58).

The *frontend* is not yet using these tokens. Okay, so the problem is actually just getting tokens from the backend to the frontend since the normal Django use case is to embed these somewhere in a page where there is form. **But** we're using DRF, not templates/forms, so how to we get these tokens to the frontend making API requests?

One has to assume this is a fairly well understood issue with Django Rest Framework. Here's what the DRF docs [say on this](http://www.django-rest-framework.org/topics/ajax-csrf-cors/):

> In order to make AJAX requests, you need to include CSRF token in the HTTP header, [as described in the Django documentation](https://docs.djangoproject.com/en/stable/ref/csrf/#ajax).

### Auth and security are sort of out of scope, but if they're not we should do auth properly

Design doc expands on this, and CSRF as a threat vector does not at all make sense given the use case in the design doc. But for the purposes of argument let's set this aside. Given the "story" of the app, *if we are going to implement auth/security* then we should probably just use authentication (i.e. only privileged users get to make special requests with `TokenAuthentication`) and set the auth token locally in an env var or something.

DRF [docs on token authentication](http://www.django-rest-framework.org/api-guide/authentication/#tokenauthentication) and helpful [blog post](http://kylebebak.github.io/post/django-rest-framework-auth-csrf).

Thoughts @lamby @WGierke?

**Comment by lamby** [comment]
Agree it's somewhat out of scope but turning it all off makes it somewhat un-Djangoey and possibly likely to scare people away in terms of "oh they clearly don't really care about the code quality at all."

Otherwise it seems quite quick to implement so on balance I would say we go ahead and commit this.

**Comment by isms** [comment]
Should be closed by 6dbc201f24c21c004901d43bd788694d919fecda.
